tests/rptest/tests/e2e_shadow_indexing_test.py

```diff
@@ -1552,3 +1563,155 @@ def test_hydration_completes_when_consumer_killed(self):
+            cloud_storage_segment_max_upload_interval_sec=1,
+        )
+        super().__init__(
+            test_context,
+            node_prealloc_count=1,
+            extra_rp_conf={
+                # Avoid segment merging so we can generate many segments
+                # quickly.
+                "cloud_storage_enable_segment_merging": False,
+                'log_segment_size_min': 1024,
+                'cloud_storage_cache_chunk_size': self.chunk_size,
+                'cloud_storage_cluster_metadata_upload_interval_ms': 1000,
+                'enable_cluster_metadata_upload_loop': True,
+                'controller_snapshot_max_age_sec': 1,
+            },
+            environment={'__REDPANDA_TOPIC_REC_DL_CHECK_MILLIS': 5000},
+            si_settings=si_settings,
+            # These tests write many objects; set a higher scrub timeout.
+            cloud_storage_scrub_timeout_s=120)
+        self.kafka_tools = KafkaCliTools(self.redpanda)
+        self.log_searcher = LogSearchLocal(test_context, [],
+                                           self.redpanda.logger,
+                                           self.redpanda.STDOUT_STDERR_CAPTURE)
+
+    def setUp(self):
+        pass
+
+    def _setUp(self, topology):
+        topology.enable_traffic_shaping()
+        topology.start_with_topology(self.redpanda)
+        for topic in self.topics:
+            self.kafka_tools.create_topic(topic)
+        rpk = RpkTool(self.redpanda)
+        rpk.alter_topic_config(self.topic,
+                               TopicSpec.PROPERTY_RETENTION_LOCAL_TARGET_MS,
+                               '1000')
+
+    @skip_debug_mode
+    @cluster(num_nodes=2)
+    def test_many_partitions_recovery_with_traffic_shaping(self):
+        """
+        Test that reproduces an OOM when doing recovery with a large dataset in
+        the bucket.
+        """
+        producer = KgoVerifierProducer(self.test_context,
+                                       self.redpanda,
+                                       self.topic,
+                                       msg_size=1024,
+                                       msg_count=10 * 1000 * 1000,
+                                       rate_limit_bps=256 *
+                                       self.small_segment_size,
+                                       custom_node=self.preallocated_nodes)
+
+        s3_service_ep = self.redpanda.si_settings.cloud_storage_api_endpoint
+
+        #  block all outgoing https if we can't get an IP address or the api endpoint is not config'ed
+        s3_service_dest = ":443"
+        try:
+            s3_service_dest = socket.gethostbyname(s3_service_ep)
+        except:
+            pass
+
+        s3_service_node = FakeNode(s3_service_dest)
+
+        def count_log_message(node, msg):
+            return len(
+                [s for s in self.log_searcher._capture_log(node, f"'{msg}'")])
+
+        with ClusterTopology() as topology:
```

Comment:
i think we may not need the whole topology setup here, it should be enough to add timeouts using a raw `NodeQdisc` ? Also to match incoming trafic you may consider adding this to the `NodeQdisc`:
```python
    def drop_incoming(self, src_address):
        self._tc_execute(
            ['qdisc', 'add', 'dev', self.dev, 'handle', 'ffff:', 'ingress'])
        self._tc_execute([
            'filter', 'add', 'dev', self.dev, 'parent', 'ffff:', 'protocol',
            'ip', 'u32', 'match', 'ip', 'src', src_address, "action", "drop"
        ])
```